### PR TITLE
Allowed for alternate way of getting Unicode character categories in  IsLikelyWordForming

### PR DESCRIPTION
**Pull request by tombogle**
Also added unit tests for this method

---
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1202)

### CODE REVIEW

File: SIL.Core.Tests/Extensions/StringExtensionTests.cs [hunk 1]
```diff
@@ -716,6 +720,79 @@ public void GetLongestUsefulCommonSubstring_SurrogatePairsWithSymbolsAsWordBreak
...
+				s == vietnameseAlternateReadingMark ? UnicodeCategory.SpacingCombiningMark :
+				CharUnicodeInfo.GetUnicodeCategory(s, i);
+
+			try
+			{
+				Assert.IsTrue(vietnameseAlternateReadingMark.IsLikelyWordForming(0, false));
+			}
+			finally
+			{
+				StringExtensions.AltImplGetUnicodeCategory = null;
```
Comment (ermshiperete) [inline SIL.Core.Tests/Extensions/StringExtensionTests.cs:792]: We could also do this in a setup method that runs before each test. That way we wouldn't have to check for `StringExtensions.AltImplGetUnicodeCategory` being `null` at the start of each test, and we wouldn't need this try/finally block here.

**Comment by tombogle** [comment]
*[`SIL.Core/Extensions/StringExtensions.cs` line 604 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1202#-N7XaEgI8lTQ-nFidttL-r1-604:-N7aST8m-oGlzdUIwiZ5:b4xwmgy) ([raw file](https://github.com/sillsdev/libpalaso/blob/003f2ede1a0fd475402ab6fac4eaba9012cb4892/SIL.Core/Extensions/StringExtensions.cs#L604)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

It might be good to add an example in the comment how you'd use ICU to get the character category, or add a unit test that demonstrates this.
</blockquote></details>

I did one better. I created and hooked up the only implementation anyone will _probably_ ever need as part of the Sldr initialization..

**Review by tombogle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1202)* status: 0 of 8 files reviewed, 5 unresolved discussions (waiting on @ermshiperete, @JohnThomson, and @papeh)

___
*[`SIL.Core.Tests/Extensions/StringExtensionTests.cs` line 775 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1202#-N7XaEgI8lTQ-nFidttK-r1-775:-N7aZjVmDgbzlaPfy_di:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/003f2ede1a0fd475402ab6fac4eaba9012cb4892/SIL.Core.Tests/Extensions/StringExtensionTests.cs#L775)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

I find this test hard to understand - it's not very clear what we currently get (without the AltImplGetUnicodeCategory extension) and what we expect (ok, it says it in the test name...).

So I wonder if it'd be better to assert that `IsLikelyWordForming` will return the expected result (which will always fail) and add a `Ignore` attribute to the test with the reasoning for this test.
</blockquote></details>

Done.
___
*[`SIL.Core.Tests/Extensions/StringExtensionTests.cs` line 782 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1202#-N7XaEgI8lTQ-nFidttK-r1-782:-N7aUtmt7LS9EC__ojv3:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/003f2ede1a0fd475402ab6fac4eaba9012cb4892/SIL.Core.Tests/Extensions/StringExtensionTests.cs#L782)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Renaming the variables to `str` and `index` would improve the documentation.
</blockquote></details>

Done.
___
*[`SIL.Core.Tests/Extensions/StringExtensionTests.cs` line 783 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1202#-N7XaEgI8lTQ-nFidttK-r1-783:-N7aVcpc7QLcyCz6mqgH:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/003f2ede1a0fd475402ab6fac4eaba9012cb4892/SIL.Core.Tests/Extensions/StringExtensionTests.cs#L783)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

To improve the documentation purpose of this test, we should compare a substring:

```suggestion
				s.Substring(i, vietnameseAlternateReadingMark.Length)  == vietnameseAlternateReadingMark ? UnicodeCategory.SpacingCombiningMark :
```
</blockquote></details>

Done.
___
*[`SIL.Core.Tests/Extensions/StringExtensionTests.cs` line 792 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1202#-N7XaEgI8lTQ-nFidttK-r1-792:-N7aWZ1GEVSMFXGiiOS6:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/003f2ede1a0fd475402ab6fac4eaba9012cb4892/SIL.Core.Tests/Extensions/StringExtensionTests.cs#L792)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

We could also do this in a setup method that runs before each test. That way we wouldn't have to check for `StringExtensions.AltImplGetUnicodeCategory` being `null` at the start of each test, and we wouldn't need this try/finally block here.
</blockquote></details>

Done.

File: SIL.WritingSystems/IcuUCharCategoryExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace SIL.WritingSystems
+{
+	public static class IcuUCharCategoryExtensions
+	{
+		public static UnicodeCategory ToUnicodeCategory(this Character.UCharCategory category)
+		{
+			switch (category)
+			{
+				case Character.UCharCategory.UNASSIGNED:
```
Comment (ermshiperete) [inline SIL.WritingSystems/IcuUCharCategoryExtensions.cs:12]: Too bad that `Character.UCharCategory` and `UnicodeCategory` don't have the same values...

**Review by ermshiperete — APPROVED** [review]
:lgtm:

**Comment by tombogle** [comment]
*[`SIL.WritingSystems/IcuUCharCategoryExtensions.cs` line 12 at r2](https://reviewable.io/reviews/sillsdev/libpalaso/1202#-N7aRofVkdEbHukzxaFV-r2-12:-N7b1FgqEEQJ5rfH8epB:bqog80t) ([raw file](https://github.com/sillsdev/libpalaso/blob/fec56aa306df156a4f06a628e15f1e1efff75320/SIL.WritingSystems/IcuUCharCategoryExtensions.cs#L12)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Too bad that `Character.UCharCategory` and `UnicodeCategory` don't have the same values...
</blockquote></details>

Agreed

**Pull request merged**